src/client/executor.rs

```diff
@@ -530,82 +523,79 @@ impl Client {
-                            session.advance_operation_time(ts);
-                        }
-                        self.update_cluster_time(&r, session).await;
-                        if r.is_success() {
-                            // Retrieve recovery token from successful response.
-                            Client::update_recovery_token(is_sharded, &r, session).await;
-
-                            Ok(CommandResult {
-                                raw: response,
-                                deserialized: r.into_body(),
-                            })
-                        } else {
-                            // if command was ok: 0, try to deserialize the command error.
-                            // if that fails, return a generic error.
-                            Err(response
-                                .body::<CommandErrorBody>()
-                                .map(|error_response| error_response.into())
-                                .unwrap_or_else(|e| {
-                                    Error::from(ErrorKind::InvalidResponse {
-                                        message: format!(
-                                            "error deserializing command error: {}",
-                                            e
-                                        ),
-                                    })
-                                }))
+                async fn handle_response<T: Operation>(
+                    client: &Client,
+                    op: &T,
+                    session: &mut Option<&mut ClientSession>,
+                    is_sharded: bool,
+                    response: RawCommandResponse,
+                ) -> Result<RawCommandResponse> {
+                    let raw_doc = RawDocument::new(response.as_bytes())?;
+
+                    let ok = match raw_doc.get("ok")? {
+                        Some(b) => crate::bson_util::get_int_raw(b).ok_or_else(|| {
+                            ErrorKind::InvalidResponse {
+                                message: format!(
+                                    "expected ok value to be a number, instead got {:?}",
+                                    b
+                                ),
+                            }
+                        })?,
+                        None => {
+                            return Err(ErrorKind::InvalidResponse {
+                                message: "missing 'ok' value in response".to_string(),
+                            }
+                            .into())
                         }
+                    };
+
+                    let cluster_time: Option<ClusterTime> = raw_doc
+                        .get("$clusterTime")?
+                        .and_then(RawBson::as_document)
+                        .map(|d| bson::from_slice(d.as_bytes()))
+                        .transpose()?;
+
+                    let at_cluster_time = op.extract_at_cluster_time(raw_doc)?;
```

Comment:
Unfortunately, the `atClusterTime` field is not always in the same spot (`find` and `aggregate` return it within the `cursor` subdocument, `distinct` has it as a top level field), so we need to introduce this new operation method to handle that.